File: 206. Reverse Linked List/solution.go
package main

func reverseList(head *ListNode) *ListNode {
	return recursive(nil, head)
}

func iterative(current *ListNode) *ListNode {
	var prev, next *ListNode

	for current != nil {
		next = current.Next
		current.Next = prev
		prev = current
		current = next
	}

	return prev
}

func recursive(prev, current *ListNode) *ListNode {
	if current == nil {
		return current
	} else if current.Next == nil {
		current.Next = prev
		return current
	}

	next := current.Next
	current.Next = prev

	return recursive(current, next)
}

File: 206. Reverse Linked List/Solution.java
// 2
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
public class Solution {
    public ListNode reverseListIteratively(ListNode head) {
        ListNode prev = null, next;

        while (head != null) {
            next = head.next;
            head.next = prev;
            prev = head;
            head = next;
        }

        return prev;
    }

    public ListNode reverseList(ListNode head) {
        return recursive(null, head);
    }

    public ListNode recursive(ListNode prev, ListNode current) {
        if (current == null) {
            return current;
        } else if (current.next == null) { // exit case
            current.next = prev;
            return current;
        }

        ListNode next = current.next;
        current.next = prev;

        return recursive(current, next);
    }
}